Mathlib/Tactic/Ring/Compare.lean

```diff
@@ -0,0 +1,215 @@
+/-
+Copyright (c) 2024 Heather Macbeth. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Heather Macbeth
+-/
+import Mathlib.Tactic.Ring.Basic
+import Mathlib.Tactic.NormNum.Ineq
+
+/-!
+# Automation for proving inequalities in commutative (semi)rings
+
+This file provides automation for proving certain kinds of inequalities in commutative semirings:
+goals of the form `A ≤ B` and `A < B` for which the ring-normal forms of `A` and `B` differ by a
+nonnegative (resp. positive) constant.
+
+For example, `⊢ x + 3 + y < y + x + 4` is in scope because the normal forms of the LHS and RHS are,
+respectively, `3 + (x + y)` and `4 + (x + y)`, which differ by a constant.
+
+## Main declarations
+
+* `Mathlib.Tactic.Ring.proveLE`: prove goals of the form `A ≤ B` (subject to the scope constraints
+  described)
+* `Mathlib.Tactic.Ring.proveLT`: prove goals of the form `A < B` (subject to the scope constraints
+  described)
+
+## Implementation notes
+
+The automation is provided in the `MetaM` monad; that is, these functions are not user-facing. It
+would not be hard to provide user-facing versions (see the test file), but the scope of this
+automation is rather specialized and might be confusing to users. It is also subsumed by `linarith`.
+-/
+
+namespace Mathlib.Tactic.Ring
+
+open Lean Qq Meta
+
+/-- Inductive type carrying the two kinds of errors which can arise in the metaprograms
+`Mathlib.Tactic.Ring.evalLE` and `Mathlib.Tactic.Ring.evalLT`. -/
+inductive ExceptType | tooSmall | notComparable
+export ExceptType (tooSmall notComparable)
+
+/-- For use in `Mathlib.Tactic.Ring.proveLE` (to short-circuit typeclass inference):
+`OrderedCommSemiring` implies `CommSemiring`. -/
+abbrev tc₁ (α : Type*) [OrderedCommSemiring α] : CommSemiring α := inferInstance
+
+/-- For use in `Mathlib.Tactic.Ring.proveLE` (to short-circuit typeclass inference):
+`OrderedCommSemiring` implies `AddMonoidWithOne`. -/
+abbrev tc₂ (α : Type*) [OrderedCommSemiring α] : AddMonoidWithOne α := inferInstance
+
+/-- For use in `Mathlib.Tactic.Ring.proveLE` (to short-circuit typeclass inference):
+`OrderedCommSemiring` implies `LE`. -/
+abbrev tc₃ (α : Type*) [OrderedCommSemiring α] : LE α := inferInstance
+
+theorem add_le_add_right' {α : Type*} [OrderedCommSemiring α] {b c : α} (bc : b ≤ c) (a : α) :
```

Comment:
I think there should be a comment explaining those lemmas have different type class assumption as their root version. I guess they also aim to be faster to apply, but I don’t see how this is relevant since their instance implicit arguments will actually be provided explicitly if I understand correctly. If I *don’t* understand correctly then maybe some comment could be useful.

Reply:
I wrote some explanation in 6c1725d421eb1982fac3d06205910df70fa29b33.

>  their instance implicit arguments will actually be provided explicitly

I don't understand this part of your comment; could you please clarify?

Reply:
I mean my understanding is those instances are never synthesized, they are used explicitly in lines like https://github.com/leanprover-community/mathlib4/pull/16840/files#diff-cea92a6714c5e25cbc74994e707df7ab77fc813dec41aec0dfdbda7d7e400b3aR118.

Reply:
When you use `q(...)` to build an expression involving a lemma like `add_le_add_right`, you don't provide the instance-implicit arguments -- instead some Qq magic in the tactic code finds these arguments (in `q(...)` form) in the context.

Does that address the question you're asking, or are you asking about something else?